allow pg query param syntax

abeidahmed opened the pull request:
### Summary

Previously, 
```ruby
ActiveRecord::DatabaseConfigurations::UrlConfig.new(:production, :production, 'postgres:///?user=user&password=passwd&dbname=theapp_production', {}).configuration_hash
``` 
was returning `{ :user=>"user", :dbname=>"theapp_production", :adapter=>"postgresql" }`. The `password` attribute was returning `nil` because `uri.password` was `nil` and merging the hashes resulted in the above result. 

This PR fixes it and now returns `{ :user=>"user", :password=>"passwd", :dbname=>"theapp_production", :adapter=>"postgresql" }`.

Issue #42797 mentions this. It also states that `postgres://user:passwd@/theapp_production` is a valid Postgres URI. Currently, if you run 
```ruby
ActiveRecord::DatabaseConfigurations::UrlConfig.new(:production, :production, 'postgres://user:passwd@/theapp_production', {})
```
it throws an error `/home/abeid/.rbenv/versions/3.0.1/lib/ruby/3.0.0/uri/generic.rb:207:in 'initialize': the scheme postgres does not accept registry part: user:passwd@ (or bad hostname?) (URI::InvalidURIError)` because the URI is not a valid [RFC 2396](https://datatracker.ietf.org/doc/html/rfc2396) implementation. Any ideas on how to fix this?

Thank you.

federicoaldunate commented on activerecord/lib/active_record/database_configurations/connection_url_resolver.rb:
```diff
@@ -67,14 +67,16 @@ def raw_config
               database: database_from_path,
               host: uri.hostname
-            )
+            }.reject! { |key, value| value.nil? }
```
What do you think? or using `reject! { |key, value| value.nil? }` is faster?
```suggestion
            }.compact!
```

abeidahmed replied:
I like the `compact!`. Let me run some benchmarks and I'll post the changes.

Navdevl commented:
@abeidahmed Adding in the benchmarks if incase needed.

**Sample code used:**
```ruby
require 'benchmark'

values = [nil, 'something']

record = {}
5000.times do |n|
  record[n] = values.sample
end

n = 100000

Benchmark.bm do |x|
  x.report { n.times do; record.reject! { |key, value| value.nil? }; end }
  x.report { n.times do; record.compact!; end }
end
```

**The output is:**

```ruby
     user     system      total        real
12.511272   0.006561  12.517833 ( 12.527778)
 2.243834   0.001086   2.244920 (  2.246199)

```

gabriel-amaral commented:
@abeidahmed Regarding the Postgres connection string not following the [RFC2396](https://datatracker.ietf.org/doc/html/rfc2396), 

Here [PostgreSQLAdapter doc](https://api.rubyonrails.org/classes/ActiveRecord/ConnectionAdapters/PostgreSQLAdapter.html ), It is already documented that there are default values for the host attribute.

I also checked, and the [ruby-pg](https://github.com/ged/ruby-pg) accepts the non-URI Postgresql connection strings (which makes sense, but I decided to check anyway).

Sample code used:
```ruby
require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  gem "rails", github: "rails/rails", branch: "main"
  gem "pg"
end

connection_str = PG::Connection.parse_connect_args('postgresql://vagrant@/postgres')
connection = PG.connect(connection_str)
puts connection.conninfo_hash.slice(:user, :password, :port, :database, :host, :dbname)
result = connection.exec('CREATE TABLE users ( id serial PRIMARY KEY, name VARCHAR(50) );')
puts result.inspect
result = connection.exec("INSERT INTO users (name) VALUES ('John Doe 1');")
puts result.inspect
result = connection.exec("SELECT * FROM users;")
puts result.map(&:to_h)

connection_str = PG::Connection.parse_connect_args('postgresql://vagrant@/test')
connection = PG.connect(connection_str)
puts connection.conninfo_hash.slice(:user, :password, :port, :database, :host, :dbname)
result = connection.exec('CREATE TABLE users ( id serial PRIMARY KEY, name VARCHAR(50) );')
puts result.inspect
result = connection.exec("INSERT INTO users (name) VALUES ('John Doe 2');")
puts result.inspect
result = connection.exec("SELECT * FROM users;")
puts result.map(&:to_h)
```

So, for me, there are few options:

- create a separated parse for Postgres' connection strings and skip the URI parse since it doesn't follow RFC.
- skip the URI parse or rescue the error and build the connection hash later based on `connection.conninfo_hash`. It doesn't feel a good solution for me though.
- Other options could be to document that the `ConnectionUrlResolver` only supports connection strings that follow the [RFC2396](https://datatracker.ietf.org/doc/html/rfc2396). In this case, it will not support all kinds of Postgres connection strings.

p8 commented:
Thanks for the PR @abeidahmed!
Can you add a test?

abeidahmed commented:
> @abeidahmed Regarding the Postgres connection string not following the [RFC2396](https://datatracker.ietf.org/doc/html/rfc2396),
> 
> Here [PostgreSQLAdapter doc](https://api.rubyonrails.org/classes/ActiveRecord/ConnectionAdapters/PostgreSQLAdapter.html), It is already documented that there are default values for the host attribute.
> 
> I also checked, and the [ruby-pg](https://github.com/ged/ruby-pg) accepts the non-URI Postgresql connection strings (which makes sense, but I decided to check anyway).
> 
> Sample code used:
> 
> ```ruby
> require "bundler/inline"
> 
> gemfile(true) do
>   source "https://rubygems.org"
> 
>   git_source(:github) { |repo| "https://github.com/#{repo}.git" }
> 
>   gem "rails", github: "rails/rails", branch: "main"
>   gem "pg"
> end
> 
> connection_str = PG::Connection.parse_connect_args('postgresql://vagrant@/postgres')
> connection = PG.connect(connection_str)
> puts connection.conninfo_hash.slice(:user, :password, :port, :database, :host, :dbname)
> result = connection.exec('CREATE TABLE users ( id serial PRIMARY KEY, name VARCHAR(50) );')
> puts result.inspect
> result = connection.exec("INSERT INTO users (name) VALUES ('John Doe 1');")
> puts result.inspect
> result = connection.exec("SELECT * FROM users;")
> puts result.map(&:to_h)
> 
> connection_str = PG::Connection.parse_connect_args('postgresql://vagrant@/test')
> connection = PG.connect(connection_str)
> puts connection.conninfo_hash.slice(:user, :password, :port, :database, :host, :dbname)
> result = connection.exec('CREATE TABLE users ( id serial PRIMARY KEY, name VARCHAR(50) );')
> puts result.inspect
> result = connection.exec("INSERT INTO users (name) VALUES ('John Doe 2');")
> puts result.inspect
> result = connection.exec("SELECT * FROM users;")
> puts result.map(&:to_h)
> ```
> 
> So, for me, there are few options:
> 
> * create a separated parse for Postgres' connection strings and skip the URI parse since it doesn't follow RFC.
> * skip the URI parse or rescue the error and build the connection hash later based on `connection.conninfo_hash`. It doesn't feel a good solution for me though.
> * Other options could be to document that the `ConnectionUrlResolver` only supports connection strings that follow the [RFC2396](https://datatracker.ietf.org/doc/html/rfc2396). In this case, it will not support all kinds of Postgres connection strings.

This would be helpful. Thank you. Let me give this one more shot!

abeidahmed commented:
Hey @p8, I added some tests. Could you please take a look and let me know? Thanks!

abeidahmed commented:
Friendly bump.

p8 commented on activerecord/lib/active_record/database_configurations/connection_url_resolver.rb:
```diff
@@ -67,14 +67,16 @@ def raw_config
-            )
+            }.compact_blank
+
+            query_hash.merge(connection_attrs)
```
I think this would be nice case for using reverse_merge:
https://edgeapi.rubyonrails.org/classes/Hash.h. tml#method-i-reverse_merge
It is used in several place in the Rails code base. 
Than we could probably skip the compact_blank.

abeidahmed replied:
```ruby
connection_attrs = {
  adapter: @adapter,
  username: uri.user, 
  password: uri.password,
  port: uri.port,
  database: database_from_path,
  host: uri.hostname
}.compact_blank

connection_attrs.reverse_merge(query_hash)
```

Did you mean something like this? Without `compact_blank` it isn't working, because `reverse_merge` doesn't reject `nil` values.

p8 replied:
Nice! I wasn’t sure it would work 😄

abeidahmed replied:
Same. I first commented that it didn't work but later I realized that I just had to replace `merge` with `reverse_merge`. Silly of me 😁

p8 commented:
@abeidahmed this looks good to me. Could you also add a Changelog entry? Then we can wrap this up.

abeidahmed committed:
Allow pg query param syntax

Allow configuring Postgres password through the socket URL.

For example:
```ruby
  ActiveRecord::DatabaseConfigurations::UrlConfig.new(
    :production, :production, 'postgres:///?user=user&password=secret&dbname=app', {}
  ).configuration_hash
```

will now return,

```ruby
  { :user=>"user", :password=>"secret", :dbname=>"app", :adapter=>"postgresql" }
```

abeidahmed commented:
Is there anything left to do in this PR? If not, can we merge this? Thanks.

p8 commented on activerecord/CHANGELOG.md:
```diff
@@ -32,6 +32,24 @@
 
     *Kevin Newton*
 
+*   Previously, some `configuration_hash`'s keys for `ActiveRecord::DatabaseConfigurations::UrlConfig`
```
Sorry, on the move this week. 
I think the title could be a bit more succinct. What do you think about:
```suggestion
*   Allow configuring Postgres password through the socket URL.
```

abeidahmed replied:
Yes, this is more clearer as to what this PR achieves. I've made changes to the CHANGELOG.md. Thanks for the help.
